### PR TITLE
describegpt: split run_inference_options into per-phase helpers

### PR DESCRIPTION
**Pull request by jqnatividad**
## Summary

Continuation of the `describegpt` decomposition work (follows merged #3720 / #3721). The ~875-line `run_inference_options` is now a thin orchestrator over eight focused helpers:

- `build_inference_messages` (was a nested `get_messages` closure) — promoted to file-scope free fn.
- `run_dictionary_phase`, `run_description_phase`, `run_tags_phase` — one per inference kind.
- `run_prompt_phase` — handles session loading, relevance-check, sliding-window, LLM call, and session-state mutation. Returns a `PromptPhaseOutcome` struct carrying the completion, `has_sql_query` flag, mutated session state, and the system prompt that the SQL phase reuses.
- `execute_sql_query_phase` — extracts the ~400-line SQL-results block (file validation, fence extraction, scoresql retry loop, DuckDB/Polars dispatch, session tracking).
- `finalize_structured_output` — JSON/TOON emission at end of run.

`run_inference_options` is now: create client, `check_model`, dispatch phases in order, apply max-tokens gate against the last phase's tokens, route SQL, finalize output, persist session.

## Minor consolidations

- `get_prompt_file(args)?.base_url` is now computed once at the orchestrator instead of 4 times (once per phase). One visible behavioral delta: a malformed prompt-file now errors **before** the first LLM call rather than after the Dictionary phase — strictly better UX (no tokens wasted on failure).
- `get_output_format(args)?` is computed once and passed to each phase instead of re-derived per phase.

## Test plan

- [x] `cargo build --locked --bin qsv -F all_features`
- [x] `cargo clippy -F all_features --bin qsv`
- [x] `cargo t describegpt -F all_features` — 64 integration tests pass
- [x] `cargo test --locked -F all_features --bin qsv describegpt::tests` — 13 unit tests pass
- [ ] Manual smoke: run `describegpt --all --prompt "..." --sql-results out.csv` against a sample CSV to exercise Dictionary → Description → Tags → Prompt → SQL pipeline end-to-end.

## Deferred (from the handoff plan)

Still open: `run` decomposition (~700 lines), `cache.rs` submodule extraction (gated on `#[io_cached]` proc-macro requalification), Tier 3 polish bundle (triplicated dictionary formatters, SQL-fence regex hoist, named constants, `.clone()` trims, parser unit tests), `try_fix_json` edge cases.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/cmd/describegpt.rs [hunk 1]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
+    #[cfg(not(feature = "polars"))]
+    {
+        if cache_type != &CacheType::Fresh && cache_type != &CacheType::None {
+            let _ = invalidate_cache_entry(args, PromptType::Prompt);
+        }
+        fail_clierror!(
+            "Cannot answer the prompt using just Summary Statistics & Frequency Distribution \
+             data. However, \"SQL RAG\" mode is only supported when the `polars` feature is \
+             enabled, or when using DuckDB via the QSV_DUCKDB_PATH environment variable."
         )
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs]: Fixed in 022b6f823. Changed the guard from `cache_type != &Fresh && cache_type != &None` to just `cache_type != &None` at all four SQL-error sites plus the non-polars fallthrough. `CacheType::Fresh` writes the completion into the underlying Disk/Redis cache (see the second `get_*_completion` call in `get_cached_completion` and the `Forces fresh API call but still updates cache` comment on the enum), so skipping invalidation on Fresh was leaving bad SQL-producing Prompt responses in cache for the next non-fresh run. This was pre-existing behavior that the review made visible; bundling the one-line fix here since it's exactly what you pointed at.

File: src/cmd/describegpt.rs [hunk 2]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
+    );
+
+    // Extract SQL query code block using regex
+    let Some(mut sql_query) = regex_oncelock!(r"(?s)```sql\s*\n(.*?)\n\s*```")
+        .captures(&completion_response.response)
+        .and_then(|caps| caps.get(1).map(|m| m.as_str().trim().to_string()))
+    else {
+        // Invalidate the prompt cache entry so user can try again without reinferring
+        // the dictionary
+        if cache_type != &CacheType::Fresh && cache_type != &CacheType::None {
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs]: Fixed in 022b6f823. Changed the guard from `cache_type != &Fresh && cache_type != &None` to just `cache_type != &None` at all four SQL-error sites plus the non-polars fallthrough. `CacheType::Fresh` writes the completion into the underlying Disk/Redis cache (see the second `get_*_completion` call in `get_cached_completion` and the `Forces fresh API call but still updates cache` comment on the enum), so skipping invalidation on Fresh was leaving bad SQL-producing Prompt responses in cache for the next non-fresh run. This was pre-existing behavior that the review made visible; bundling the one-line fix here since it's exactly what you pointed at.

File: src/cmd/describegpt.rs [hunk 3]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
+                    if stderr.to_ascii_lowercase().contains(" error:") {
+                        track_sql_error_in_session(
+                            session_state.as_mut(),
+                            session_path_owned.as_ref(),
+                            format!("DuckDB SQL query execution failed: {stderr}"),
+                        );
+                        if cache_type != &CacheType::Fresh && cache_type != &CacheType::None {
+                            let _ = invalidate_cache_entry(args, PromptType::Prompt);
+                        }
+                        return fail_clierror!("DuckDB SQL query execution failed: {stderr}");
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs:3247]: Fixed in 022b6f823. Changed the guard from `cache_type != &Fresh && cache_type != &None` to just `cache_type != &None` at all four SQL-error sites plus the non-polars fallthrough. `CacheType::Fresh` writes the completion into the underlying Disk/Redis cache (see the second `get_*_completion` call in `get_cached_completion` and the `Forces fresh API call but still updates cache` comment on the enum), so skipping invalidation on Fresh was leaving bad SQL-producing Prompt responses in cache for the next non-fresh run. This was pre-existing behavior that the review made visible; bundling the one-line fix here since it's exactly what you pointed at.

File: src/cmd/describegpt.rs [hunk 4]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
+                Err(e) => {
+                    track_sql_error_in_session(
+                        session_state.as_mut(),
+                        session_path_owned.as_ref(),
+                        format!("DuckDB SQL query execution failed: {e}"),
+                    );
+                    if cache_type != &CacheType::Fresh && cache_type != &CacheType::None {
+                        let _ = invalidate_cache_entry(args, PromptType::Prompt);
+                    }
+                    return Err(e);
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs:3260]: Fixed in 022b6f823. Changed the guard from `cache_type != &Fresh && cache_type != &None` to just `cache_type != &None` at all four SQL-error sites plus the non-polars fallthrough. `CacheType::Fresh` writes the completion into the underlying Disk/Redis cache (see the second `get_*_completion` call in `get_cached_completion` and the `Forces fresh API call but still updates cache` comment on the enum), so skipping invalidation on Fresh was leaving bad SQL-producing Prompt responses in cache for the next non-fresh run. This was pre-existing behavior that the review made visible; bundling the one-line fix here since it's exactly what you pointed at.

File: src/cmd/describegpt.rs [hunk 5]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
+    }
 
-            // Add SQL results if available (for refinement context)
-            if is_refinement && let Some(ref results) = session.sql_results {
-                messages.push(json!({
-                        "role": "assistant",
-                        "content": format!("Here are the first 10 rows from the last successful SQL query execution:\n\n```csv\n{results}\n```")
-                    }));
-            }
+    json!(messages)
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs]: Fixed in 022b6f823. Replaced `json!(messages)` with `serde_json::Value::Array(messages)` — direct constructor instead of re-processing the already-built `Vec<serde_json::Value>` through the json! macro.

File: src/cmd/describegpt.rs [hunk 6]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
+    // dictionary so the prompt phase can reuse it without re-inferring.
+    let dictionary_cache_type = if args.flag_prompt.is_some() && args.flag_fresh {
+        if args.flag_redis_cache {
+            &CacheType::Redis
+        } else {
+            &CacheType::Disk
+        }
+    } else {
+        cache_type
+    };
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs]: Fixed in dd35118e1. Added `cache_type != &CacheType::None` to the override condition so `--no-cache --prompt X --fresh` keeps caching disabled for the dictionary phase too. Pre-existing behavior before this refactor; bundled here since the review made it visible.

File: src/cmd/describegpt.rs [hunk 7]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
+    }
+
+    let sql_query_start = Instant::now();
+    print_status(
+        &format!(
+            "\nSQL results file specified.\n  Executing SQL query and saving results to \
+             {sql_results}..."
+        ),
+        None,
+    );
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs]: Fixed in dd35118e1. Rewrote the writability probe as a loop over `[sql_results_path, sql_results_path.with_extension("csv")]` so the Polars `.csv` sibling is caught if it's already on disk and read-only. DuckDB writes to the primary path so the extra probe is a harmless no-op there. Didn't modify the status message for the execution pass since both engines start writing at `{sql_results}` — the Polars rename happens on success.

File: src/cmd/describegpt.rs [hunk 8]
```diff
@@ -2565,874 +2565,1000 @@ fn process_phase_output(
...
-            completion_response.token_usage.completion,
+            last_completion.token_usage.completion,
         );
     }
 
     print_status("LLM inference/s completed.", Some(llm_start.elapsed()));
 
-    // if args.flag_output is set, display the output to the console
     if let Some(output) = &args.flag_output {
         print_status(&format!("Output written to {output}"), None);
```
[the remark this answers is not in the document]
Comment (jqnatividad) [reply src/cmd/describegpt.rs]: Fixed in dd35118e1. TSV runs now get `"TSV output written using {output} as the base path (one derived file per phase, e.g. {filestem}.{kind}.tsv)"` while JSON/TOON/Markdown keep the original `Output written to {output}`. Dispatch is a single `if output_format == OutputFormat::Tsv` check where the status fires.
